### ISSUE TITLE
Add the ability to also have additional templates on a local machine

### ISSUE THREAD

**Issue by mrlacey** [microsoft/TemplateStudio#919]
Allow the configuration of a folder on the local machine that can contain templates which will also be combined with the templates downloaded by the wizard.

The primary aim of this is to make it easy for people to have their own templates that they can use but which wouldn't make sense or be possible to include in the official list.  
That this can also help when creating new "official" templates is a bonus.

Functionality needed:
- [ ] Ability to configure a path to use via a setting within VS. This needs to support the same accessibility and localization requirements as the rest of the wizard.
- [ ] Need to save the path and have some basic validation on it.
- [ ] Path can be on the same drive as the extension&VS, on a different one or on a network share
- [ ] Only going to support a single path
- [ ] Official templates will take priority and be loaded first and then any extra local ones. (Won't have support for suppressing the official templates.)
- [ ] Need to test for and have appropriate handling for clashes in names, IDs, etc. between templates
- [ ] Need robust error handling for issues with loading extra templates and useful, detailed error messages. It might be useful to report any found errors loading templates but allow the wizard to continue with the ones it can load.
- [ ] It might be useful to have a template (& config) validation tool available to help with template authoring. We could also run this against the official templates as part of automated testing too.
- [ ] Should support adding extra pages, features, frameworks, and project types.
- [ ] Need some sample extra templates for testing.
- [ ] Will need to improve documentation on creating templates.
- [ ] This will also need to work with the VsEmulator
- [ ] Will need automated testing to verify this works
- [ ] No version checking will be done for locally loaded templates. We'll just assume they're compatible and if not give appropriate errors.

Other functionality will build upon this change adding a place for settings and the ability to include additional templates.

**Comment by crutkas** [comment]
isn't this part of #234 ?

**Comment by mrlacey** [comment]
Yes, this is a dupe of #234 (I couldn't find that before so created this.)

Closing a duplicate of #234

**Comment by mrlacey** [comment]
Actually not a dupe. This is about having templates on the local machine.
#234 is about having an additional remote source of templates, this is about having a local copy of additional templates. Similar but different.

**Comment by crutkas** [comment]
was thinking about this and #234 and i think one big thing we need to have is force custom feeds to have a min and target.  If you're past the target / tested version for that feed's author, it could create a bad template.

Local is less of an issue but for custom, i think this is a must and i would imagine the two features are going to be hand and hand for most code

**Comment by mrlacey** [comment]
If the wizard and official feed are always targeting the last 2 versions of the platform then forcing other feeds to be compatible with the extensions version should make this a non-issue.

Or did you mean a min & target version of the extension/templates that are supported?

Even if not, we need to think carefully about how we support older versions of feeds.
- At the moment we require version matching (major & minor parts) of the templates and extension.
- We don't want other feeds to suddenly break every time the extension is [automamtically] updated because we release a new version.
- We don't want people to disable auto-update of the extension as they'd miss out on bug fixes and enhancements.
- We (ideally) don't want to have to support templates targeting older versions of the extension as this would mean we couldn't introduce breaking changes to the templates. (Note that 1.2 introduced breaking changes and I expect more over the next few releases, especially, but not limited to, when needing to support solutions with multiple projects.)

So what to do? My current thinking is:
- Attempt to load templates with a version that matches the same version as the extension.
- Attempt to load templates with a version number that is one version older than the extension and also log an information message that old templates have been loaded and they may have compatibility issues. We'll need to improve the testing, handling, and logging of incompatible templates.
- Do not attempt to load templates with a version number two or more versions older than the extension. (Log a warning that any such templates were not included.)
- Use a major version update for anything that could be a breaking change. (This is good practice anyway. #SemVer)
- If we can. (External pressures and requests allowing. 😉) Defer adding this functionality until some more of the bigger planned changes are in and we can be more confident about the stability of the current template system.

**Comment by crutkas** [comment]
I am thinking more in terms of WTS.  If we do a breaking change, I don't want to be in the biz of supporting external stuff.

**Comment by mrlacey** [comment]
Absolutely.
If we disable their stuff on every update we'll make it hard for people to trust they can use us.
But no, we don't want to take responsibility for supporting or fixing their stuff.

I think the above is a good compromise of trying to support them when we make a (hopefully) non-breaking change but being clear about it when we do make a breaking change.
We probably also need announcement channels for upcoming releases and especially if they contain a potentially breaking change.

**Comment by mrlacey** [comment]
Pushing to 1.5 as non-essential; not sure it'll happen quickly; and want more experience from supporting the creation of more templates first (which hope to get from Caliburn, Prism, & VB work)

**Comment by mrlacey** [comment]
Bumping to a later milestone as it's potentially too big to bring in at the same time as VB support. Only doing one big change to supported templates at a time is probably best.

**Comment by mrlacey** [comment]
Removing my assignment and moving to the backlog as this is very unlikely to happen for a long time. Major template structure changes are planned for things like multi-project support and these will have an impact on this feature. It's going to be much easier to do this when the template structure is stable or it'll risk lots of issues for people trying to create their own templates as the knock-on effect of handling support for them.

This issue was raised as I created a crude version of this functionality for my own testing purpose in the run-up to the release of version 1.0. A version of that code was made public in branch [MattPlayground-DO-NOT-MERGE](https://github.com/Microsoft/WindowsTemplateStudio/tree/MattPlayground-DO-NOT-MERGE) but that is now very out of date as apart from updating from master when the branch was made public it hasn't really been touched since the beginning of May. That branch may or may not be useful when it comes time to implement this feature.

**Comment by mrlacey** [comment]
Doing this will require updates to the composition filters of existing templates so that they are not picked up by any new templates. For example, if wanting to add a new project type but with an existing framework, this will pick up lots of composition items which currently only filter on framework because that's sufficient for the current templates.

Also affects #234 and may affect Xamarin (#321 ?) templates too.

(issue discovered while trying to implement #1336)

**Comment by mrlacey** [comment]
In theory, this will be easy to do once #234 is done. That will download templates and copy them locally. Once the wizard can handle templates downloaded from multiple URIs it should be simple to also take the templates from a local directory or network share (assuming they can be treated the same.)

We just need a way to configure the directory (or directories) to look in. This is covered in the POC mentioned in #234 and in the "Matts Playground" branch.

**Comment by mrlacey** [comment]
For **versioning**, different versions of templates, intended for different Wizard versions can just be kept in different folders and the referenced paths updated by the user.

For **Template identities**, we should encourage naming conventions that will make them unique but WTS should ignore any 3rd party templates that have names which match official template names.

**Dependencies** need to be possible so that anyone can add a template for a page or feature that behaves and has the same functionality as an official template.

**Composition** is the potentially most complicated issue here. We need a 3rd party template to be able to add a page (or feature) that behaves and is treated like an official one. We also want them to be able to be totally separate if needed. Some experimentation and investigation in this area is definitely needed but my instinct is to include all composition templates by default but add the ability to opt-out of specific ones on a named basis, but with wildcard matching. (e.g. `"noComposition":["wts.Page.*"]` to exclude the inclusion of any composition template with a name starting "wts.Page." that would have otherwise been included.)

3rd party templates must be able to extend the project type and framework **catalogs** as a key use case is businesses with their own project type templates on a local network share. Without this functionality, WTS cannot be used for these scenarios without forking the source.  
Local templates are also an excellent way to develop and test templates for distribution in other ways. This includes the creation of templates to support new project types and frameworks.

**Comment by sibille** [comment]
Ok, I can see your point with extending the **catalog**. We'll have to study a way how catalog's from different sources can be merged and used to extend the WTS catalog.

**Comment by mrlacey** [comment]
Due to a lack of wider interest in this issue and other changes being planned for the way the extension will work in VS2022 (& beyond) it does not feel appropriate to keep this issue open as wider changes make this broadly irrelevant.
